target/scripts/helper-functions.sh

```diff
@@ -288,27 +288,65 @@ export -f _monitored_files_checksums
 
+# Outputs the DNS label count (delimited by `.`) for the given input string.
+# Useful for determining an FQDN like `mail.example.com` (3), vs `example.com` (2).
+function _get_label_length
+{
+  local INPUT=${1}
+  awk -F '.' '{ print NF }' <<< "${INPUT}"
+}
+
+# Sets HOSTNAME and DOMAINNAME globals used throughout the scripts,
+# and any subprocesses called that intereact with it.
 function _obtain_hostname_and_domainname
 {
-  if [[ -n "${OVERRIDE_HOSTNAME}" ]]
+  # Normally this value would match the output of `hostname` which mirrors `/proc/sys/kernel/hostname`,
+  # However for legacy reasons, the system ENV `HOSTNAME` was replaced here with `hostname -f` instead.
+  #
+  # TODO: Consider changing to `DMS_FQDN`; a more accurate name, and removing the `export`, assuming no
+  # subprocess like postconf would be called that would need access to the same value via `$HOSTNAME` ENV.
+  #
+  # TODO: `OVERRIDE_HOSTNAME` was introduced for non-Docker runtimes that could not configure an explicit hostname.
+  # k8s was the particular runtime in 2017. This does not update `/etc/hosts` or other locations, thus risking
+  # inconsistency with expected behaviour. Investigate if it's safe to remove support. (--net=host also uses this as a workaround)
+  export HOSTNAME="${OVERRIDE_HOSTNAME:-"$(hostname -f)"}"
+
+  # If the container is misconfigured.. `hostname -f` (which derives it's return value from `/etc/hosts` or DNS query),
+  # will result in an error that returns an empty value. This warrants a panic.
+  if [[ -z ${HOSTNAME} ]]
   then
-    export HOSTNAME="${OVERRIDE_HOSTNAME}"
-    export DOMAINNAME="${DOMAINNAME:-${HOSTNAME#*.}}"
-    # Handle situations where the hostname is name.tld and hostname -d ends up just showing "tld"
-    if [[ ! "${DOMAINNAME}" =~ .*\..* ]]
-    then
-      DOMAINNAME="${HOSTNAME}"
-    fi
-  else
-    # These hostname commands will fail with "hostname: Name or service not known"
-    # if the hostname is not valid (important for tests)
-    HOSTNAME="$(hostname -f)"
-    DOMAINNAME="${DOMAINNAME:-$(hostname -d)}"
-    if [[ ! "${DOMAINNAME}" =~ .*\..* ]]
+    dms_panic__misconfigured 'obtain_hostname' '/etc/hosts'
+  fi
+
+  # If the `HOSTNAME` is more than 2 labels long (eg: mail.example.com),
+  # We take the FQDN from it, minus the 1st label (aka _short hostname_, `hostname -s`).
+  #
+  # TODO: For some reason we're explicitly separating out a domain name from our FQDN,
+  # `hostname -d` was probably not the correct command for this intention either.
+  # Needs further investigation for relevance, and if `/etc/hosts` is important for consumers
+  # of this variable or if a more deterministic approach with `cut` should be relied on.
+  if [[ $(_get_label_length "${HOSTNAME}") -gt 2 ]]
+  then
+    if [[ -n ${OVERRIDE_HOSTNAME} ]]
     then
-      DOMAINNAME="${HOSTNAME}"
+      # Emulates the intended behaviour of `hostname -d`:
+      DOMAINNAME="$(echo "${HOSTNAME}" | cut -d '.' -f2-99)"
```

Comment:
No "99" limit:

```suggestion
      DOMAINNAME="$(echo "${HOSTNAME}" | cut -d '.' -f2-)"
```

Reply:
Thanks, I just carried that over from the original PR, would it also be helpful here to add an inline comment that it's taking everything after the 1st label? Or do you have a better approach instead of `cut` for this?

I don't know how many maintainers are as familiar/comfortable with lines like this, I know I definitely didn't know much shell syntax when I started helping maintain the project, and different utilities like `cut` parameters would confuse me even more so (_although I haven't tried to look up if `-f2-` would make sense to me reading docs, I know I could probably find `-f<number>` at least_).

Reply:
First: I am never against comments, even if I don't would add them myself in some cases. `cut` is heavily used in this project, so the usage should be clear/familiar. There is also no prevention, to look up unknown things in the man pages or do a web search 😉

>  I know I definitely didn't know much shell syntax

And every time you stumble upon something like this, you learn something new. Same here. When I first saw this (in another code), I didn't know the meaning. Now I do and use it myself 👍 So this is never an excuse for me, not to use something. Just because someone wouldn't understand it at first..

> Or do you have a better approach instead of cut for this?

You asked for it, you get it 😆 

`DOMAINNAME=${HOSTNAME#*.}`

Explained:

```
#     remove string from the beginning (non-greedy) --> ## (greedy)
*     wildcard
.     dot

--> Remove all strings beginning from the left, until (including) the first dot.
```

But someone not familiar with bash variable expansion will "complain" about this too. This solution is more efficient (no sub-process needed). I admit, more readable on **first** sight, is probably the `cut` command.
I would use the variable expansion.

Reply:
Ah awesome :)

You make some good points so I'm fine with whatever you advise :+1: 

Cheers for all the insights!

Reply:
Take the variable expansion 👍

Reply:
If it means anything: It took me half an hour to figure out what the heck `#*.}` did... I highly highly recommend against anything that wouldn't be obvious to anyone of any bash skill level.

Reply:
> If it means anything: It took me half an hour to figure out what the heck `#*.}` did...

Does it help that I added a comment about it in the commit, and that anyone who uses `git blame` should be able to arrive on this PR and come across this review sub-thread about it?

Reply:
> > If it means anything: It took me half an hour to figure out what the heck `#*.}` did...
> 
> 
> 
> Does it help that I added a comment about it in the commit, and that anyone who uses `git blame` should be able to arrive on this PR and come across this review sub-thread about it?

Sure, the comment is fine. 👍🏼Just seems like overkill/clutter.